Config option to disable hooks

fmuyassarov opened the pull request:
Fixes: https://github.com/kubernetes-sigs/node-feature-discovery/issues/859

marquiz commented on source/local/local.go:
```diff
@@ -133,6 +168,8 @@ func parseFeatures(lines [][]byte) map[string]string {
 
 // Run all hooks and get features
 func getFeaturesFromHooks() (map[string]string, error) {
+	klog.Warning("hooks are DEPRECATED since v0.12.0; use feature files instead")
```
Yeah, it will be v0.12.0 so it's correct. We can keep it here or drop it whatever you like.

Unrelated: maybe you could supplement this with a mention that they will be dropped
```suggestion
	klog.Warning("hooks are DEPRECATED since v0.12.0 and support will be removed in a future release; use feature files instead")
```

fmuyassarov replied:
Thanks, updated.

marquiz commented:
ping @eero-t @mythi

eero-t commented on docs/advanced/customization-guide.md:
```diff
@@ -153,6 +153,16 @@ should be placed in a separate directory in order to avoid NFD unnecessarily
 trying to execute them. A subdirectory under the hooks directory can be used,
 for example `/etc/kubernetes/node-feature-discovery/source.d/conf/`.
 
+**NOTE:** Hooks are being DEPRECATED and will be removed in the future releases.
```
For consistency with other changes, use "in a future release."

fmuyassarov replied:
done

eero-t commented on docs/advanced/worker-configuration-reference.md:
```diff
@@ -329,6 +329,23 @@ sources:
+
+Configuration option to disable/enable hooks execution. Enabled by default.
+Hooks are DEPRECATED since v0.12.0 release and support will be removed in a
+future release. Use [feature files](./customization-guide.md#feature-files)
```
This could include links to the relevant tickets, so that interested parties can read the discussion:
* Step 1, add option: https://github.com/kubernetes-sigs/node-feature-discovery/issues/859
* Step 2, disable by default: https://github.com/kubernetes-sigs/node-feature-discovery/issues/855
* Step 3, remove hook support: https://github.com/kubernetes-sigs/node-feature-discovery/issues/856

fmuyassarov replied:
done

eero-t commented on source/local/local.go:
```diff
@@ -133,6 +168,8 @@ func parseFeatures(lines [][]byte) map[string]string {
 
 // Run all hooks and get features
 func getFeaturesFromHooks() (map[string]string, error) {
+	klog.Warning("hooks are DEPRECATED since v0.12.0 and support will be removed in a future release; use feature files instead")
```
If this is logged every time hooks are called (which happens periodically), it could be too much.

IMHO it would be better to log the warning only at start, when reading local config, if HooksEnabled is set.

@marquiz?

(Otherwise looks OK to me.)

fmuyassarov replied:
I agree on this point. Initially I added in https://github.com/kubernetes-sigs/node-feature-discovery/blob/3df81d1ace1f74b43a74f998f8b238df6a53ca6a/pkg/nfd-client/worker/nfd-worker.go#L156 but then moved to here. Would https://github.com/kubernetes-sigs/node-feature-discovery/blob/3df81d1ace1f74b43a74f998f8b238df6a53ca6a/pkg/nfd-client/worker/nfd-worker.go#L156 be okay ?

fmuyassarov replied:
it would basically log once, as general warning regardless user enables/disabled the hooks.

eero-t replied:
@marquiz should really be the one commenting on where to do the check...

(I filed the ticket for hooks disabling + few other things, so Markus CC'd me to this. However, this is first time I'm looking at NFD code, so I'm not at all familiar how it works.)

marquiz replied:
Thanks @eero-t for the review.

@fmuyassarov I would probably put the log message later in this function if some hook files are really found and executed 🤔 Better not to print a warning for people who don't use hooks (most of them, I think). We could ratelimit into only one message e.g. with a boolean flag (that could be reseted e.g. in `SetConfig()` of the local source.

fmuyassarov replied:
Sure thing @marquiz . So we have two options I guess...
**1.** in `SetConfig()`, 

```
// SetConfig method of the LabelSource interface
func (s *localSource) SetConfig(conf source.Config) {
	switch v := conf.(type) {
	case *Config:
		s.config = v
		if s.config.HooksEnabled {
			klog.Warning("hooks are DEPRECATED since v0.12.0 and support will be removed in a future release; use feature files instead")
		}
	default:
		klog.Fatalf("invalid config type: %T", conf)
	}
}
```
log trace when hooks enabled
```
I0826 07:54:40.275016       1 nfd-worker.go:154] Node Feature Discovery Worker v0.12.0-devel-76-ga3100ce-dirty
I0826 07:54:40.275237       1 nfd-worker.go:155] NodeName: 'nfd-worker'
I0826 07:54:40.275465       1 nfd-worker.go:422] configuration file "/etc/kubernetes/node-feature-discovery/nfd-worker.conf" parsed
W0826 07:54:40.275535       1 local.go:80] hooks are DEPRECATED since v0.12.0 and support will be removed in a future release; use feature files instead
I0826 07:54:40.275547       1 nfd-worker.go:460] worker (re-)configuration successfully completed
I0826 07:54:40.275592       1 base.go:127] connecting to nfd-master at nfd-master:8080 ...
I0826 07:54:40.275630       1 component.go:36] [core]parsed scheme: ""
I0826 07:54:40.275640       1 component.go:36] [core]scheme "" not registered, fallback to default scheme

```
log trace when hooks enabled
```
I0826 07:58:39.551162       1 nfd-worker.go:154] Node Feature Discovery Worker v0.12.0-devel-76-ga3100ce-dirty
I0826 07:58:39.551321       1 nfd-worker.go:155] NodeName: 'nfd-worker'
I0826 07:58:39.552185       1 nfd-worker.go:422] configuration file "/etc/kubernetes/node-feature-discovery/nfd-worker.conf" parsed
I0826 07:58:39.552341       1 nfd-worker.go:460] worker (re-)configuration successfully completed
I0826 07:58:39.552453       1 base.go:127] connecting to nfd-master at nfd-master:8080 ...
I0826 07:58:39.552612       1 component.go:36] [core]parsed scheme: ""
I0826 07:58:39.552627       1 component.go:36] [core]scheme "" not registered, fallback to default scheme
```

**2.** in `getFeaturesFromHooks()`. When hooks are enabled, depreciation message gets printed on every reload
```
func getFeaturesFromHooks(s *localSource) (map[string]string, error) {
.....
	if s.config.HooksEnabled {
		klog.Warning("hooks are DEPRECATED since v0.12.0 and support will be removed in a future release; use feature files instead")
	}
}
```
log trace when hooks enabled
```

I0826 07:42:56.606694       1 local.go:116] starting hooks...
W0826 07:42:56.618914       1 local.go:203] hooks are DEPRECATED since v0.12.0 and support will be removed in a future release; use feature files instead
I0826 07:42:56.645574       1 nfd-worker.go:471] starting feature discovery...
I0826 07:42:56.646019       1 nfd-worker.go:483] feature discovery completed
I0826 07:42:56.646139       1 nfd-worker.go:564] sending labeling request to nfd-master
E0826 07:43:56.700575       1 kernel.go:132] failed to read kconfig: failed to read kernel config from [ /proc/config.gz /host-usr/src/linux-5.15.0-46-generic/.config /host-usr/src/linux/.config /host-usr/lib/modules/5.15.0-46-generic/config /host-usr/lib/ostree-boot/config-5.15.0-46-generic /host-usr/lib/kernel/config-5.15.0-46-generic /host-usr/src/linux-headers-5.15.0-46-generic/.config /lib/modules/5.15.0-46-generic/build/.config /host-boot/config-5.15.0-46-generic]
I0826 07:43:56.702207       1 local.go:116] starting hooks...
W0826 07:43:56.709296       1 local.go:203] hooks are DEPRECATED since v0.12.0 and support will be removed in a future release; use feature files instead
I0826 07:43:56.735445       1 nfd-worker.go:471] starting feature discovery...
I0826 07:43:56.735863       1 nfd-worker.go:483] feature discovery completed
I0826 07:43:56.735929       1 nfd-worker.go:564] sending labeling request to nfd-master
```
log trace when hooks disabled
```
I0826 07:44:05.110774       1 nfd-worker.go:211] reloading configuration
I0826 07:44:05.116216       1 nfd-worker.go:422] configuration file "/etc/kubernetes/node-feature-discovery/nfd-worker.conf" parsed
I0826 07:44:05.116438       1 nfd-worker.go:460] worker (re-)configuration successfully completed
E0826 07:44:05.119196       1 kernel.go:132] failed to read kconfig: failed to read kernel config from [ /proc/config.gz /host-usr/src/linux-5.15.0-46-generic/.config /host-usr/src/linux/.config /host-usr/lib/modules/5.15.0-46-generic/config /host-usr/lib/ostree-boot/config-5.15.0-46-generic /host-usr/lib/kernel/config-5.15.0-46-generic /host-usr/src/linux-headers-5.15.0-46-generic/.config /lib/modules/5.15.0-46-generic/build/.config /host-boot/config-5.15.0-46-generic]
I0826 07:44:05.145666       1 nfd-worker.go:471] starting feature discovery...
I0826 07:44:05.146036       1 nfd-worker.go:483] feature discovery completed
I0826 07:44:05.146066       1 nfd-worker.go:564] sending labeling request to nfd-master
```

 I personally prefer option 1 but fine with other opinions too. WDYT?

fmuyassarov replied:
Perhaps it is better to keep it as it is since the logs get printed only when hooks are enabled? It is true that users might not be using it although enabled and still see the logs, but it seems fine to me because we warning them about the feature that they should not enable. WDYT @marquiz ?

marquiz replied:
In my thinking this might create confusion if you get warning like this with the default configuration (and you're not using hooks). But maybe it's good to print anyway - but in this case I'd probably do some rate limiting as I suggested, e.g. once after every config change or smth. WDYT?

marquiz reviewed (changes requested):
Thanks @fmuyassarov for this. Looks good, almost ready for merging. In addition to addressing the comment on logging, please run `make templates` so that the example config in the Helm chart also gets updated.

/ok-to-test

fmuyassarov commented:
> Thanks @fmuyassarov for this. Looks good, almost ready for merging. In addition to addressing the comment on logging, please run `make templates` so that the example config in the Helm chart also gets updated.
> 
> /ok-to-test

Thanks, done.

marquiz commented on source/local/local.go:
```diff
@@ -143,6 +183,9 @@ func getFeaturesFromHooks() (map[string]string, error) {
 		}
 		return features, fmt.Errorf("unable to access %v: %v", hookDir, err)
 	}
+	if !(len(files) > 0) {
+		printHooksWarning = false
+	}
```
Hmm, now it looks that we only print the warning when NO hooks are found 🧐

I'd just print it here (without any ratelimiting)

```suggestion
	if !(len(files) > 0) {
		klog.Warning("hooks are DEPRECATED since v0.12.0 and support will be removed in a future release; use feature files instead")
		}
```

fmuyassarov replied:
@marquiz No, I would say it is the opposite
`if !(len(files) > 0)` -> when hooks dir is empty, don't print. Perhaps you missed exclamation mark

eero-t replied:
That should really be written without '!' for clarity: `if len(files) == 0`.

Also, above warning is about *use* of a deprecated API.

Because enabling hooks opens potential NFD security hole [1], I think security warning should be shown when hooks are enabled.  For example:
`klog.Warning("insecure hooks API enabled with 'HooksEnabled: true' (DEPRECATED since v0.12.0)")`

[1] of NFD executing random binaries from host, which may be  insecure, or even potentially malicious, and hooks require having extra components in NFD container which may later turn out to contain security holes.

fmuyassarov replied:
I agree with your first point on simplifying condition. But for the warning message it seems we are overcomplicating it. There could be another issue apart from security. I think, if there are some drawbacks of using a feature it should be documented clearly but not necessarily warned in the log trace.

marquiz replied:
> of NFD executing random binaries from host

I wouldn't say random binaries from host. System / cluster admin is responsible for those in a similar way that any "system" files, I'd say. If the administrator just randomly installs any scripts or runs random containers with root privileges nfd is one of the minor problems :)

eero-t replied:
They are random from the **NFD** PoV, and may force keeping it less secure.  It does not e.g. know whether what they do matches NFD security context settings: can some of its security context settings tightened or would that break some random hook?

marquiz replied:
I get that but nfd-worker (with the default deployments) runs with probably the tightest security context possible: fileystem and all mounts are ro, run-as-non-root, allow-privilege-escalation=false, all caps dropped...

eero-t replied:
Ok. I concur that warning about deprecated API *use* is enough.  And while there's no rate-limiting, it's indeed better just to output that warning directly in getFeaturesFromHooks().   Adding package-global "printHooksWarning" flag just to do warning from calling function is redundant complication.

fmuyassarov replied:
@eero-t No problem. @marquiz I think it is good to merge it now.

marquiz replied:
Yeah, looks good now

marquiz reviewed (approved):
Thanks @fmuyassarov, good to merge from my point of view

marquiz commented:
> Because enabling hooks opens potential NFD security hole [1], I think warning should be shown (also) when hooks are enabled.

I think we could print another note (possibly just an info/note, not a warning at this stage when `hooksEnabled=true` by default) in `SetConfig`. This would be printed whenever config is changed. Can be done in a separate PR, too

fmuyassarov commented:
/cc @zvonkok @ArangoGutierrez

zvonkok commented:
On PTO will take a look next week.

ArangoGutierrez commented:
I like this PR
but since it introduces a deprecating change, let's go for more eyes
cc @zvonkok @yevgeny-shnaidman

fmuyassarov commented:
/cc @Ethyling for lgtm :)

zvonkok commented:
/lgtm
